src/meta/src/stream/scale.rs

```diff
@@ -221,6 +221,17 @@ where
     ) -> MetaResult<()> {
         let ctx = self.build_reschedule_context(&reschedule).await?;
 
+        let mut revert_funcs = scopeguard::guard(
+            vec![],
+            |revert_funcs: Vec<futures::future::BoxFuture<()>>| {
+                tokio::spawn(async move {
+                    for revert_func in revert_funcs.into_iter().rev() {
+                        revert_func.await;
+                    }
+                });
+            },
+        );
```

Comment:
I'm not sure whether this is the best practice for reverting something... Running revert funcs with `tokio::spawn` gives no insurance on whether or when it will succeed, which may break the consistency of the barrier manager as we make it more complicated.

cc @zwang28

Reply:
> I'm not sure whether this is the best practice for reverting something

I totally agree we want a better pattern. @BugenZhao 

The revert funcs was initially used to simplify code where there are many Err exits and one Ok exit, and we know those revert funcs are run with best effort. https://github.com/risingwavelabs/risingwave/blob/b969abea6db29da7edb88114598e9e59cd67c8ca/src/meta/src/stream/stream_manager.rs#L356

Then when applied the same pattern to place where there are multiple Ok exit, it begins to look weird because we need to call revert_funcs.clear multiple times. (Fortunately there are only 2 here)
https://github.com/risingwavelabs/risingwave/blob/b969abea6db29da7edb88114598e9e59cd67c8ca/src/meta/src/stream/source_manager.rs#L590

Reply:
So is there a better way to roll back here? Like golang's defer, we can remove it if it's not used in the right way, the logic here is not very complicated, I thought this was the recommended practice when I saw it used elsewhere

Reply:
Added a wrapper methods to make revert func executed in the same task.
https://github.com/risingwavelabs/risingwave/pull/5148